Detectors/DCS/include/DetectorsDCS/DCSProcessor.h

```diff
@@ -0,0 +1,143 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef DETECTOR_DCS_DCSPROCESSOR_H_
+#define DETECTOR_DCS_DCSPROCESSOR_H_
+
+#include <memory>
+#include <Rtypes.h>
+#include <unordered_map>
+#include <deque>
+#include "Framework/Logger.h"
+#include "DetectorsDCS/DataPointCompositeObject.h"
+#include "DetectorsDCS/DataPointIdentifier.h"
+#include "DetectorsDCS/DataPointValue.h"
+#include "DetectorsDCS/DeliveryType.h"
+
+/// @brief Class to process DCS data points
+
+namespace o2
+{
+namespace dcs
+{
+
+class DCSProcessor
+{
+
+  using Ints = std::vector<int>;
+  using Chars = std::vector<char>;
+  using Doubles = std::vector<double>;
+  using Binaries = std::array<uint64_t, 7>;
+  using Strings = std::array<char, 56>;
+
+  using DQChars = std::deque<char>;
+  using DQInts = std::deque<int>;
+  using DQDoubles = std::deque<double>;
+  using DQUInts = std::deque<uint32_t>;
+  using DQBools = std::deque<bool>;
+  using DQStrings = std::deque<Strings>;
+  using DQTimes = std::deque<uint32_t>;
+  using DQBinaries = std::deque<Binaries>;
+
+  using DPID = o2::dcs::DataPointIdentifier;
+  using DPVAL = o2::dcs::DataPointValue;
+  using DPCOM = o2::dcs::DataPointCompositeObject;
+
+ public:
+  DCSProcessor() = default;
+  ~DCSProcessor() = default;
+
+  void init(const std::vector<DPID>& aliaseschars, const std::vector<DPID>& aliasesints,
+            const std::vector<DPID>& aliasesdoubles, const std::vector<DPID>& aliasesUints,
+            const std::vector<DPID>& aliasesbools, const std::vector<DPID>& aliasesstrings,
+            const std::vector<DPID>& aliasestimes, const std::vector<DPID>& aliasesbinaries);
+
+  void init(const std::vector<DPID>& aliases);
+
+  int process(const std::unordered_map<DPID, DPVAL>& map);
+
+  template<typename T>
+  int processArrayType(const std::vector<DPID>& array, DeliveryType type, const std::unordered_map<DPID, DPVAL>& map, std::vector<uint64_t>& latestTimeStamp, std::unordered_map<DPID, T>& destmap);
```

Comment:
The template definition should be in the header file, either in the class declaration body or after it, as inline methods are usually defined, i.e.
````
template<typename T>
int DCSProcessor::processArrayType(....) {
   // processing the array of type T
   int found = 0;
   // implementation you have currently in cxx file 
}
````

Reply:
Not sure I understand the comment: I declared the specialized templated function in the header, but the body is in the cxx.

Reply:
no, you have the non-specialized template implementation in the cxx, while it should be ``.h``

Reply:
Ah, the non-specialized!
But are you sure it is necessary? Like it is now, it works (compiles + macro works).

Reply:
In general, it is necessary, in your case it works because at the moment these methods are used only in the same compilation module where they are defined. But you will not be able to call them as class methods from some another module.

Reply:
Hello Ruben,

Changed according to our discussion.

Chiara